lib/utils/deeply-non-duplicated-addon.js

```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+/**
+ * Deduplicate one addon's children addons recursively from hostAddons.
+ *
+ * @private
+ * @param {Object} hostAddons
+ * @param {EmberAddon} dedupedAddon
+ * @param {String} treeName
+ */
+module.exports = function deeplyNonDuplicatedAddon(hostAddons, dedupedAddon, treeName) {
+  if (dedupedAddon.addons.length === 0) {
+    return;
+  }
+
+  dedupedAddon._orginalAddons = dedupedAddon.addons;
+  dedupedAddon.addons = dedupedAddon.addons.filter(addon => {
+    if (addon.lazyLoading && addon.lazyLoading.enabled) {
+      return true;
+    }
+
+    if (addon.addons.length > 0) {
+      addon._orginalAddons = addon.addons;
+      deeplyNonDuplicatedAddon(hostAddons, addon, treeName);
```

Comment:
As implemented, it appears (although it is possible I have misread) we only prevent duplication when a `hostAddon` is included in one of the descendent addons. I believe this is actually incomplete, as duplicate descendent addons should also be handled.

Here is a more concrete example:

Given the following add-on dependency graph:

```
a -> b
a -> c
a -> d
b -> d 
d -> c
```

Assuming none of these are `hostAddons` I believe we will still duplicate `d` (via `a` and `b`) twice, and `c` (via `a`, `b`, `d`) three times.

Reply:
good gotcha @stefanpenner I had not realized that.

Reply:
For your example, if none of these are `hostAddons`, all of these will be `lazy engines addons`, so the situation is kinda like a normal ember app without engines.
Do you think the normal ember app will have the duplication in it's "vendor.js" if all of these addons are `app addons`? I think the anwser is no, I might find some realted code of "ember-cli", https://github.com/ember-cli/ember-cli/blob/master/lib/broccoli/merge-trees.js#L60 and https://github.com/ember-cli/ember-cli/blob/master/lib/models/addon.js#L635 and https://github.com/ember-cli/ember-cli/blob/master/lib/models/addon.js#L654.

So the duplication this PR aim to address is between the host app and lazy engines and I think there's no duplication of addons issue within a normal app.

Reply:
@stefanpenner can you reproduce your scenario in a public repository? using https://github.com/kaliber5/ember-cli-bundle-analyzer

I think ember-engines should be concerned about this, since this use case is specific to the ecosystem of ember engines and isolation.

Reply:
I'll take a look today.

Reply:
@stefanpenner any new on this?

Reply:
I believe @2hu12 raises a good point, this likely is not needed and if it is can be brought back in at a later time.

Reply:
Is not doubling addons a critical problem? I am convinced that yes, but we need to know where this should be solved. It should be resolved here or at ember-cli.

We can not miss the tracking of this, and we can open a issue.

Reply:
@villander I just had a verification on this, please check it out here https://github.com/2hu12/ember-engine-verification. It has a lazy engine `engine-a-lazy`, and three in-repo-addon `addon-a`, `addon-b` and `addon-c`, all three in-repo-addons are all not "hostAddon", and the dependency graph is like:
```
a -> c
b -> c
```
And the result is that it only has one piece of "addon-c" in https://github.com/2hu12/ember-engine-verification/blob/master/dist/engines-dist/engine-a-lazy/assets/engine-vendor.js#L29

Reply:
@2hu12 thanks for the verification!

Reply:
@2hu12 thank you so much for explain this.

the duplicity of `hostAddons` should be a problem that the ember-cli team must deal with.

Many thanks for doing this here, ember-engines is much better now because of you.